### PR TITLE
cluster upgrade: migrate the wizard to PF4, add progress view

### PR DESCRIPTION
**Pull request by sjd78**
## Summary of Changes
There are three phases of changes.

**First**, the i18n use of the cluster upgrade modal is refactored to match the use of all of the rest of the modals and dashboard.  The component external i18n support was originally built for the wizard to be shared with other projects.  That never happened, and now that support only served to complicate the modal.  All of the i18n use was directly to where it is used in the components.

**Second**, cluster upgrade was migrated to to PF4.  The modal is the only PF3 modal remaining in the project.  PF3 use blocks easy upgrade to webpack 5 (and building in EL9).  Only a few future changes will be needed to free the project from PF3.

Old PF3 components / helpers that are no longer needed and have been removed:
  - `BaseFormGroup`
  - `FieldLevelHelp` (the PF3 version)
  - `WizardPattern`
  - `MessageDialog`
  - `SpinnerDialog`

The PF4 version of Cluster Upgrade Wizard follows the PF4 designs made awhile ago: https://marvelapp.com/prototype/d9aaf81/screen/64616225.  Some of the styles in the design have been updated in current PF4, and the current styles are used.

**Third**, add an upgrade progress tracking `isFinished` step to the wizard.  The percent complete will be determined by looking at events with the correct correlation id and extracting the progress from there. See https://github.com/oVirt/ovirt-ansible-collection/pull/415 for how the progress events are logged and their format.  After the upgrade is requested, events will be followed.  Closing the wizard from the progress tracking step will not cancel the actual upgrade.

## Demo
Demo of the new wizard in action (using the `FAKE_DATA=true` build flag):
ezgif com-gif-maker (1)

## Screen Shots
Loading modal:
screenshot-engine-dev ovirt_8080-2022 03 04-13_43_02

Cluster maintenance policy check confirmation modal:
screenshot-engine-dev ovirt_8080-2022 03 04-13_43_12

Wizard steps (Select Hosts, Upgrade Options, Review):
screenshot-engine-dev ovirt_8443-2022 03 27-11_08_49

screenshot-engine-dev ovirt_8443-2022 03 27-11_09_33

screenshot-engine-dev ovirt_8443-2022 03 27-11_09_33a

Wizard Tracking Progress (in progress and complete:
screenshot-engine-dev ovirt_8443-2022 03 28-13_50_38
screenshot-engine-dev ovirt_8443-2022 03 28-13_50_43

### CODE REVIEW

File: src/modals/cluster-upgrade/SelectHosts.js
Comment (rszwajko) [inline src/modals/cluster-upgrade/SelectHosts.js:42]: just in case  host or name is undefined please use `?.`

File: src/modals/cluster-upgrade/ClusterUpgradeModal.js
Comment (sjd78) [reply src/modals/cluster-upgrade/ClusterUpgradeModal.js:49]: There are a few attributes that are needed by the wizard variant.  I already refactored to directly specify everything and skip this style.

File: src/modals/cluster-upgrade/ClusterUpgradeModal.js
Comment (sjd78) [reply src/modals/cluster-upgrade/ClusterUpgradeModal.js:63]: In my current version (I'll push it soon) it looks ok inline.  I don't see much utility in extracting them to small components.

File: src/modals/cluster-upgrade/ClusterUpgradeWizard.js
Comment (sjd78) [reply src/modals/cluster-upgrade/ClusterUpgradeWizard.js:82]: non-issue, it's changed in my current code

File: src/modals/cluster-upgrade/SelectHosts.js
Comment (sjd78) [reply src/modals/cluster-upgrade/SelectHosts.js:42]: That won't help anything.  If those fields are missing, the entire wizard will not work.

**Comment by sgratch** [comment]
@sjd78 please remember to keep dialogs consistent with webadmin by locating the action buttons on the right corner of the dialog.

**Comment by sjd78** [comment]
Current push works properly now.  Final style adjustments are pending.  I will squash the commits and fixup the screen shots once the styles are adjusted.

**Review by sgratch — COMMENTED** [review]
@sjd78 
Style adjustments comments (based on what I tested on my env):

- the dialog title background is black instead of default color used on other dialogs.

- The action buttons are still located on the right corner of the dialog with a wrong order: next, back, cancel instead of back, next, cancel

- no tooltip/info text field added to last "Review" step for explaining that close/go to events buttons can be clicked without terminating the upgrade process.

image

**Review by rszwajko — APPROVED** [review]
Looks good. Thanks for nice "fake" data!
I have some concerns regarding self-triggering hook but this can wait until the next version.

**Comment by sjd78** [comment]
Latest push solves all of the style issues @sgratch pointed out.

I didn't change anything WRT @rszwajko's `TrackProgress` data flow comments.

Only outstanding work is to (1) fix (or remove) the simple log viewer on the `TrackProgress` component, and (2) update the screen shots and demo.

**Comment by sgratch** [comment]
@sjd78 
Regarding style issues-  looks good now.
Regarding `TrackProgress` data flow comments by Radek - worth fixing now? 
Please keep in mind that QE should verify basic functionality comparing to old dialog before merging.

File: src/modals/cluster-upgrade/TrackProgress.js
Comment (sjd78) [reply src/modals/cluster-upgrade/TrackProgress.js:51]: I don't see how doing anything more than the `useEffect()` would buy anything useful in this case.

In my first implementation of the tracking code, I had all of it happen in the `ClusterUpgradeDataProvider`.  All of the data lived there and all of the data was pushed down via props to the `TrackProgress` component.  It did work from a data flow perspective, but there were UI render issues.  Keeping it in the `TrackProgress` component avoid the render issues.

**Comment by sjd78** [comment]
> @sjd78 Regarding style issues- looks good now. 

:+1:

> Regarding `TrackProgress` data flow comments by Radek - worth fixing now? 

IDK.  I'll triple check his comments, but at some point what really is the point of changing working code between different styles?

> Please fix build errors. 

Done.

> Need to keep in mind that QE should verify basic functionality comparing to old dialog before merging.

Right. I'll run OST and ping Ivana.

File: src/modals/cluster-upgrade/TrackProgress.js
Comment (sjd78) [reply src/modals/cluster-upgrade/TrackProgress.js:64]: > The usual flow is that the effect is triggered by some data. In this case we have `correlationId`. As long as the trigger remains the same the effect is not be executed. Here the effect is executed once but then it self-triggers internally. My concerns:
> 
>   1. canceling the effect when the component is unmounted - we cancel the timer but if the fetch is in progress then it will set the state  anyway which will cause errors. (btw DataProviderHook has most likely similar problem)
> 

Yep, we have no provisions to handle canceling the fetch if the component is unmounted in this effect or in any data provider.  Given the limited scope of the modals in general, I don't see it as much of a problem.

>   2. fetch error will make it stop working - the user will probably just close the dialog but it's not ideal

That's a case that does make sense to handle.  On fetch error, skip calling the tick and just schedule the next one.

>   3. progress of progress - the refresh rate is fixed but the fetch make take a while  - it would be nice to have some indicator that the work is in progress

Well the progress bar is blue indicating the process in ongoing.  Lame I know.  There isn't really a nice place to add a spinner since it would just be running until the tracker sets state 'complete'.  I can add one if there is a good place to add one.

>   4. tight coupling of "data provider" and "scheduler". IMHO separating both would reduce complexity. Additionally we could use standard provider and just trigger it periodically i.e.
> 
> ```js
> const [trigger, setTrigger] = useState(0)
> const { data, fetchError, fetchInProgress } = useDataProvider({ fetchData, trigger})
> 
> useEffect(() => {
>   scheduleTrackTick(TIME, correlationId, ()=> setTrigger(trigger++))
>   return  () => {
>       cancelTracker(correlationId)
>     }
> }
> ```

There isn't much of a difference in how the component would look between those two forms.  I'm not convinced that the `useEffect` hook isn't sufficient.

> 
> As soon as the progress will be part of the Cluster entity we could also trigger re-resh on the main data provider.

Yeah probably, but still have the annoying blinking wizard in modal issue to deal with.

**Comment by sgratch** [comment]
@pnovotny please update once sanity/regression tests are completed for the PF4 cluster upgrade dialog. We want to merge this PR if all passed successfully.

**Comment by pnovotny** [comment]
LGTM.
Verified with build ovirt-engine-ui-extensions-1.3.1-1.20220328git70e96ea.el8.noarch.rpm and the cluster upgrade works as expected, no regressions.

**Review by sgratch — APPROVED** [review]
LGTM

**Pull request merged**
